### ISSUE TITLE
Loose Surface Form Matching

### ISSUE THREAD

**Issue by abhishekg2389** [dbpedia-spotlight/dbpedia-spotlight#363]
Hi @tgalery,

As we were discussing about loose surface form matching in https://github.com/dbpedia-spotlight/pignlproc/pull/17, you stated the following pipeline:
1) Get the spot
2) Generate new Surface Froms
3) Get all the candidates in both the uppercase SF store and the lowercase sf store
As per this pipeline I have implemented the step 2 and want to connect it to the both endpoints (Step 1 and Step 3). So can you indicate me where are these endpoints, if there are any.

**Comment by tgalery** [comment]
So basically you would have to change the `MemmorySurfaceFormStore` class (which is re-used across endpoints). We had a quick and dirty stab at a related problem here: https://github.com/dbpedia-spotlight/dbpedia-spotlight/pull/284/files .
This PR basically tries to account cases where there are more than one space between the tokens of an sf, e.g. `Barack          Obama`, but the method `getRankedSurfaceFormCandidates` is pretty much the same, it tries to find candidates based on the uppercase SF store, and then if nothing is found move to the lowercase one. You would need to change this logic as you indicated in your point 3. Let me know if you have more specific questions.

**Comment by abhishekg2389** [comment]
Hi @tgalery,

Thanks for this endpoint specification.
Additionally let me discuss plural remover (https://github.com/dbpedia-spotlight/pignlproc/pull/17#issuecomment-112745686) here, as it is concerned to this issue.
Actually I am first trying to normalize the spot, which I got in step 1, then I will generate different probable sfs based on this normalized spot.

**Comment by tgalery** [comment]
I think the more general issue is that these transformations would be language dependent. I think you can read the language from the model you pass into the main class, but this means that you'd need to write specifc transformers for each lang.

**Comment by abhishekg2389** [comment]
@tgalery 
Yes. But at least for the proof of concept sake we can use the implementation for English and later on if it works well we will implement for other languages. So currently we will use transformers for English and skip any such transformations for other languages. So for this can you please tell which method would you like me to implement.

1. Replacing with rules - "s" -> "", "ses" -> "s", "xes" -> "x", "zes" -> "z", "ches" -> "ch", "shes" -> "sh", "men" -> "man", "ies" -> "y" 
2. Use stemmer by John Carroll: https://github.com/knowitall/morpha
In option 1 we might miss some complex like men, children. But besides these rules I have a semi-exhaustive list of plurals to singular conversion.
Option 2 might be more exhaustive than in Option 1 but it will be more time consuming.

Moreover it would be great if I could get raw SF store to do some analysis regarding the probability of surface form given candidate or vice-versa.

**Comment by tgalery** [comment]
It if we are introducing a new lib, we could have to do some research on it first, and multi-language support would be a good feature. For the purposes of just get the work going, maybe you should go in the first way

What you will need to do is (i) add the plural form if the origin sf is singular and the other way around, and (ii) remove head articles and prepositions if they have one (add one if they don't).

**Comment by abhishekg2389** [comment]
Hi @tgalery,

I have coded step 2 to some extent. I have also ran my code on a sample spot. Please take a look at it. https://github.com/abhishekg2389/dbpedia-spotlight/blob/master/SpotGenerator.java
Besides this I have couple of concerns:
1) Converting a plural to singular is good but converting a singular to plural might cause some problem. For example, in "Michael Jackson" which token ("Michael", "Jackson") should I make plural. If we are considering all the combination of tokens then It might be feasible in this case (2 tokens) but when it will come to a longer spot (United Nations Educational, Scientific and Cultural Organization (UNESCO)) it might cause problem.
2) Removal of preposition/conjunction is good but while adding them same case comes like where to insert the preposition or conjunction.

**Comment by abhishekg2389** [comment]
Besides this can you provide me raw SF store to do some analysis regarding the probability of surface form given candidate or vice-versa.

**Comment by abhishekg2389** [comment]
Hi @tgalery 
Can you please check the flow so that I can proceed with further implementation.

**Comment by tgalery** [comment]
Hey @abhishekg2389 I took a look. I think you are in the right direction, but at the moment, I think this is overkill. In your example, it feels that a lot of sfs would be generated from a single input, and I'm sure this might bring a lot of noise. I think the best way to go about this would be to incrementally add the changes. That is just transform the word in a lower case version, and remove articles at the beginning and then look up in the two sf stores. This could be implemented quite easily on the scala side and we could try to test that with some datasets out there.

**Comment by abhishekg2389** [comment]
Hi @tgalery,

I am implementing the algorithm in MemmorySurfaceFormStore class. But there is one function for editDistanceScore() which I want to modify based on some statistics. I want to get those statistics from SurfaceForm store like how many time there is only initials in SF etc. The reason for this is as follows:
Suppose the Spotted Candidate is "M Jackson" and one of the Probable SF Candidate generated is "M J", but as per current implementation we are using LevenshteinDistance as the distance between these two which will probably comes out to be 6 which is quite unlike. This is because we use initials quite frequently and conversion from "Jackson" to "J" should be either 1 edit distance away or a little bit more than 1, say 1.25 or so.
To find these type of statistics I need Raw SurfaceForm store data which have mapping from SurfaceForms to Entitites. So I request you it would be great if you could get that data or point me to the link from where I can get that data, as I cannot generate that data from the code because I don't have that much computing power with me right now.

Thanks,
Abhishek

**Comment by tgalery** [comment]
Hi @abhishekg2389 we used to have this within the downloads folder of spotlight, but for some reason it's gone. I'm gonna see if I can dig this out from another spotlighter and make it available somewhere, but it might take a while. If you want you can extract this info yourself. If you use the sbt console, you can load the spotlight jar and all the classes would be available to you. You could then create an SF store and loop through the sfs and write the information to a file. As for the editDistance function, I wouldn't change that. I would write another distanceFunction and just use this new one when calculating the stats.

**Comment by abhishekg2389** [comment]
Hi @tgalery,
For SF store I am not sure If I would be able to create that because I don't have that much computation power with me. But I will try for sure If I can. As far as editDistance function is concerned I earlier meant to say what you said only.

**Comment by tgalery** [comment]
I think that the count files are back up again: you can download them here http://spotlight.sztaki.hu/downloads/raw/

**Comment by abhishekg2389** [comment]
Hi @tgalery,

I want to know if we are also recording that the Entity is a person or organization or anything else somewhere. If yes then from where can I get the data or how can I access that data.
